clients/src/main/java/org/apache/kafka/common/utils/PoisonPill.java

```diff
@@ -24,19 +24,52 @@
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import javax.management.MBeanServer;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Value;
 
 
+class PoisonPillStats {
+    private final Metrics metrics;
+    private final Sensor timeSinceLastDequeueSensor;
+    public PoisonPillStats(Metrics metrics) {
+        this.metrics = metrics;
+
+        String metricGroupName = "PoisonPill-metrics";
+        this.timeSinceLastDequeueSensor = metrics.sensor("time-since-last-dequeue");
+
+        this.timeSinceLastDequeueSensor.add(metrics.metricName("time-since-last-dequeue",
+            metricGroupName,
+            "The latest time-since-last-dequeue in milliseconds"),
+            new Value());
+        this.timeSinceLastDequeueSensor.add(metrics.metricName("time-since-last-dequeue-max",
+            metricGroupName,
+            "The maximum value of time-since-last-dequeue in milliseconds"),
+            new Max());
+    }
+
+    public void recordTimeSinceLastDequeInMs(final long timeSinceLastDequeue) {
+        this.timeSinceLastDequeueSensor.record(timeSinceLastDequeue);
+    }
+}
+
 public class PoisonPill {
+    private final PoisonPillStats poisonPillStats;
+
+    public PoisonPill(Metrics metrics) {
+        poisonPillStats = new PoisonPillStats(metrics);
+    }
 
-    public static void die() {
+    public void die() {
         die(null, -1, 1);
     }
 
-    public static void die(File heapDumpFolder, final long maxWaitForDump) {
+    public void die(File heapDumpFolder, final long maxWaitForDump) {
         die(heapDumpFolder, maxWaitForDump, 1);
     }
 
-    public static void die(File heapDumpFolder, final long maxWaitForDump, int haltStatusCode) {
+    public void die(File heapDumpFolder, final long maxWaitForDump, int haltStatusCode) {
```

Comment:
these can still be static methods

Reply:
They can be static mathods. But static methods are hard to mock, so I try to avoid them as much as possible.